### ISSUE TITLE
mouseleave event isn't fired when moving the mouse outside the window (Windows)

### ISSUE THREAD

**Issue by MartinMa** [electron/electron#611]
I have a simple `div` with a css hover effect changing the background color. When I move the mouse outside of the window area, the `div` keeps its hover style. It should actually go back to its normal style. Presumable the mouseleave event was "swallowed".

For a demonstration video see here: https://www.screenr.com/4M0N

It happens using the frameless view **and** using the windows frame view. Even though the latter is more difficult to reproduce (move the mouse faster).

You can reproduce it with this sample code:

``` html
<!DOCTYPE html>
<html>
  <head>
    <title>mouseleave</title>

        <style>
            .box {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                left: 10%;
            }
            .box:hover {
                background-color: red;
            }
        </style>
  </head>
  <body>
        <div class="box"></div>
  </body>
</html>
```

My setup:
Windows ~~**8.1**~~ **7**
atom-shell **0.15.7**

**Comment by Keavon** [comment]
This also happens when moving your mouse from an item with a hover style to an item with the `-webkit-app-region: drag;` style applied in addition to moving from an item with a hover style applied to outside the frameless window.

**Comment by KeitIG** [comment]
I can confirm what @Keavon is saying.

It also happens when moving the mouse from an :hover-ed element to a `-webkit-app-region: drag;`, the first element still get the :hover

**Comment by abnerlee** [comment]
Does anyway found solutions for this bug?

**Comment by KeitIG** [comment]
I'm afraid this is related to Chromium, so we can't really make something

**Comment by sapjax** [comment]
https://www.npmjs.com/package/electron-drag  look this

**Comment by Acenth** [comment]
This seems to happen in both frameless and non-frameless windows, with and without 'webkit-app-region'. It doesn't seem to happen in Atom IDE or Chrome. It does happen in Visual Studio code though and my test apps.

Anyone have insight into it?

**Comment by hadar111** [comment]
Reproduce in my app too :(
Anyone have insight into it?

**Comment by Acenth** [comment]
Tried a bunch different things to get it working, but it happens in every version of Chromium, including frameless Chrome apps. We submitted an issue to Chromium but they upgraded to VS2015 that same day, so I imagine they got flooded with more important bugs.

**Comment by mangopearapples** [comment]
Does anyone have a work around for this problem? I'm getting the same issue.

**Comment by madipor** [comment]
does anyone found solutions for this?

**Comment by louisameline** [comment]
I wrote a workaround for this, not ideal but it's enough for my use case. See how the window mouseleave is detected, even when leaving the app-region top bar: 

![fix](https://cloud.githubusercontent.com/assets/371814/23146371/811b8eaa-f7d6-11e6-96f4-2c40c1ef2d4d.gif)

The code is here: https://gist.github.com/louisameline/1213bb112c6cb12a98b2ab525dfb8b07

**Comment by mangopearapples** [comment]
Nice one! Perhaps you could prevent the events triggering while a window is overlaying it by checking for focus?

**Comment by louisameline** [comment]
I thought about that but then a blurred window wouldn't react to mouse hovering, even if there were no windows on top of it. That's not the usual behavior and I'm not sure it would be better.

**Comment by louisameline** [comment]
When I say the window wouldn't react, I'm actually thinking about my app-region titlebar, which wouldn't reappear because it can't catch mouse moves and needs coordinates calculations instead. But you're probably right, with appropriate modifications on this bar it may work as expected.

**Comment by louisameline** [comment]
I have enabled the check for the focused window by default, but you can disable it by setting `regionAppOptimized = true`. Most of the time, app regions will fail to respond if you don't set to true:

![fix](https://cloud.githubusercontent.com/assets/371814/23161350/16fb2892-f82a-11e6-97e8-58d9afe01445.gif)

I thought it would be smart to apply a `-webkit-app-region: drag` property on the titlebar **after** it got its opacity back (to let it catch mouseenter events), but that trick doesn't work. The view must have that info in cache and does not obey the CSS change quickly enough.

**Comment by codebytere** [comment]
Given that there exist workarounds, this is something that isn't on our current roadmap and which I am thus goin to tag as `wontfix`.

**Comment by MartinMa** [comment]
This should work out of the box. There is a single workaround suggested here, which still seems to have issues (If I understand correctly).

The code suggested by me a couple of years ago, needs to be incorporated in the message loop handler. I would suggest a pull request, but I need some help in **where to find the message loop** in the source code. Any hints?

**Comment by codebytere** [comment]
Awesome! Also happy to help you get this across the finish line; i'm marking these PRs wontfix not because we don't care, just because it's not realistic for you to hope we'll fix something that's not on any imminent horizon or todo list.

**Comment by MartinMa** [comment]
I cannot reproduce this issue with the latest releases of Electron (3.0.0 and 2.0.10) anymore. I tried two different machines (Windows 10 and Windows 7). So I guess this issue has been fixed *by accident* somewhere in between. Maybe with a new Chromium version.

In fact, the latest release where I can reproduce the issue using the hover example above is the (meanwhile) unsupported version 1.6.18. The first release where this issue seems to be fixed is 1.7.0 beta (May 10th 2017). I think this issue can just as well be marked as fixed. Any objections?

**Comment by otoomey** [comment]
Getting the same issue with electron 7.1.4. Both the css `:hover` selector and javascript `mouseout` and `mouseleave` events _sometimes_ do not fire. I am on Windows 10 Pro, using a frameless, transparent window. 

One work around is to use a mouse event capture tool like [iohook](https://wilix-team.github.io/iohook/), but I can't help but feel that this is definitely a bug (especially that the javascript event doesn't always fire).

**Comment by vifird** [comment]
Appeal to reopen this issue!

**Comment by frankhale** [comment]
I don't think they will fix it, even their Github Desktop app exhibits this unfortunate behavior. Minimize the app and then hover over the taskbar icon in Windows and you'll notice that they cannot react to the minimize state and the minimize icon is still highlighted in the preview.

Here is a screenshot showing (yeah it's hard to see but it's there)

![github-desktop-electron-mouse-event-outside-window-issue](https://user-images.githubusercontent.com/190154/71303253-41834a80-2384-11ea-8d33-4a4d02425c58.png)

**Comment by vifird** [comment]
I just found a hacky solution, which works for me. use `mainWindow.setResizable(true)`.
I had tested on Window10, with electron v6.0.2.

**Comment by SlicerNorthD** [comment]
mouseleave event doesn't fire with the below window properties set but if resizable is set to true then it works. Using windows 10 with electron 7.3.1. Any way to work around this?

alwaysOnTop: true,
skipTaskbar: true,
thickFrame: false,
frame: false,
resizable: false,
webPreferences: {
    nodeIntegration: true,
}

**Comment by wangdingwei** [comment]
> mouseleave event doesn't fire with the below window properties set but if resizable is set to true then it works. Using windows 10 with electron 7.3.1. Any way to work around this?
> 
> alwaysOnTop: true,
> skipTaskbar: true,
> thickFrame: false,
> frame: false,
> resizable: false,
> webPreferences: {
> nodeIntegration: true,
> }

same problem

**Comment by sssooonnnggg** [comment]
I add a fix for this bug, see #29721

**Comment by sssooonnnggg** [comment]
我英文不太好，尝试解释一下 bug 产生的细节：

经研究发现：

 `shell/browser/ui/views/frameless_view.cc` 里的处理有问题：

当 resizable 设置为 false 时，鼠标经过窗口边缘时（窗口边缘设置为5个像素[kResizeInsideBoundsSize]）， 调用 `FramelessView::ResizingBorderHitTest` 时返回了 HTBORDER。

这会导致 Windows 误以为鼠标离开了窗口区域，从而向窗口发送了一个 WM_MOUSELEAVE 事件，实际上鼠标依然在窗口内。

chromium 在 `legacy_render_widget_host_win.cc` 里处理 WM_MOUSELEAVE 时，会通过 `::GetCursorPos()` 和 `::WindowFromPoint()` 来获取鼠标当前所在窗口 :

```c++
LRESULT LegacyRenderWidgetHostHWND::OnMouseLeave(UINT message,
                                                 WPARAM w_param,
                                                 LPARAM l_param) {
  mouse_tracking_enabled_ = false;
  LRESULT ret = 0;
  HWND capture_window = ::GetCapture();
  if ((capture_window != GetParent()) && GetWindowEventTarget(GetParent())) {
    // We should send a WM_MOUSELEAVE to the parent window only if the mouse
    // has moved outside the bounds of the parent.
    POINT cursor_pos;
    ::GetCursorPos(&cursor_pos);

    // WindowFromPoint returns the top-most HWND. As hwnd() may not
    // respond with HTTRANSPARENT to a WM_NCHITTEST message,
    // it may be returned.
    HWND window_from_point = ::WindowFromPoint(cursor_pos);
    if (window_from_point != GetParent() &&
        (capture_window || window_from_point != hwnd())) {
      bool msg_handled = false;
      ret = GetWindowEventTarget(GetParent())->HandleMouseMessage(
          message, w_param, l_param, &msg_handled);
      SetMsgHandled(msg_handled);
    }
  }
  return ret;
}
```

此时因为 WM_NCHITTEST 返回了 HTBORDER，因此虽然触发了 WM_MOUSELEAVE，但是鼠标仍然在窗口内部。

所以会导致 `window_from_point != hwnd()` 这个条件不会满足，进而 `GetParent())->HandleMouseMessage(message, w_param, l_param, &msg_handled)` 这个函数不会被执行，从而导致 DOM 的 hover 状态不会被清除。

所以当 BrowserWindow 的 resizable 属性设置为 false 时，NCHitTest 应该返回 HTCLIENT，这样才能让 Windows 在鼠标真正离开窗口时触发 WM_MOUSELEAVE，解决 hover 状态清除不掉的问题。
